src/Transport/ResponseCorrelator.cs

```diff
@@ -85,15 +87,40 @@ public override FutureResponse AsyncRequest(Command command)
 			return future;
         }
 
-        public override Response Request(Command command, TimeSpan timeout)
+        public override Task<Response> RequestAsync(Command command, TimeSpan timeout)
         {
+	        TaskCompletionSource<Response> taskCompletionSource = new TaskCompletionSource<Response>(TaskCreationOptions.RunContinuationsAsynchronously);
+			if (timeout.TotalMilliseconds > 0)
+	        {
+		        CancellationTokenSource ct = new CancellationTokenSource(timeout);
+		        ct.Token.Register(() =>
```

Comment:
Register returns `CancellationTokenRegistration` that also should be disposed when you're dealt with this request.